bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/OAuthConnector.java

```diff
@@ -0,0 +1,383 @@
+import java.security.AccessController;
+import java.security.PrivilegedActionException;
+import java.security.PrivilegedExceptionAction;
+import java.time.LocalDateTime;
+import java.util.Base64;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.client.api.ContentResponse;
+import org.eclipse.jetty.client.api.Request;
+import org.eclipse.jetty.client.util.FormContentProvider;
+import org.eclipse.jetty.http.HttpHeader;
+import org.eclipse.jetty.http.HttpMethod;
+import org.eclipse.jetty.http.HttpStatus;
+import org.eclipse.jetty.util.Fields;
+import org.eclipse.smarthome.core.auth.client.oauth2.AccessTokenResponse;
+import org.eclipse.smarthome.core.auth.client.oauth2.OAuthException;
+import org.eclipse.smarthome.core.auth.client.oauth2.OAuthResponseException;
+import org.eclipse.smarthome.io.net.http.HttpClientFactory;
+import org.eclipse.smarthome.io.net.http.TrustManagerProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.FieldNamingPolicy;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonSyntaxException;
+
+/**
+ * Implementation of the OAuthConnector. It directly deals with the underlying http connections (using Jetty).
+ * This is meant for internal use. OAuth2client's clients should look into {@code OAuthClientService} or
+ * {@code OAuthFactory}
+ *
+ * @author Michael Bock - Initial contribution
+ * @author Gary Tse - ESH adaptation
+ *
+ */
+@NonNullByDefault
+public class OAuthConnector {
+
+    private static final String HTTP_CLIENT_CONSUMER_NAME = "OAuthConnector";
+
+    private final HttpClientFactory httpClientFactory;
+
+    private static final Logger logger = LoggerFactory.getLogger(OAuthConnector.class);
+    private final Gson gson;
+
+    public OAuthConnector(HttpClientFactory httpClientFactory) {
+        this.httpClientFactory = httpClientFactory;
+        gson = new GsonBuilder().setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES).create();
+    }
+
+    /**
+     * Authorization Code Grant
+     *
+     * @param authorizationEndpoint The end point of the authorization provider that performs authorization of the
+     *            resource owner
+     * @param clientId Client identifier (will be URL-encoded)
+     * @param redirectURI RFC 6749 section 3.1.2 (will be URL-encoded)
+     * @param state Recommended to enhance security (will be URL-encoded)
+     * @param scope Optional space separated list of scope (will be URL-encoded)
+     *
+     * @return A URL based on the authorizationEndpoint, with query parameters added.
+     * @see https://tools.ietf.org/html/rfc6749#section-4.1.1
+     */
+    public static String getAuthorizationUrl(String authorizationEndpoint, String clientId,
```

Comment:
This static methods somehow seems to be at the wrong place here - the OAuthConnector is an internal class and no API and is normally used as an instance. From what I can see, this method is only called from within `OAuthClientServiceImpl` - I'd therefore claim that moving this method to that class would be much cleaner (and your logger could be made non-static, this is actually how I noticed this here...)

Reply:
I agree it is strange.  I've modified the code to be non-static.
The original reason was that this method is entirely self-containing and refers to nothing at all.

Reply:
Why not moving it to OAuthClientServiceImpl, it this is the only code that actually NEEDs that method?